chips/nrf52840/src/ieee802154_radio.rs

```diff
@@ -767,128 +835,255 @@ impl<'a> Radio<'a> {
+                    // here to optimize the time taken to send an ACK. If the receive function was called in the
+                    // RX state handler, there is a non deterministic time required to complete the function as it may
+                    // be passed up the entirety of the networking stack (leading to ACK timeout being exceeded).
                     self.rx_client.map(|client| {
-                        let rbuf = self.rx_buf.take().unwrap(); // Unwrap fail = RX Buffer produced error when sending received packet to requestor
+                        // Unwrap fail = Radio RX Buffer is missing (may be due to receive client not replacing in receive(...) method,
+                        // or some instance in  driver taking buffer without properly replacing).
+                        let rbuf = self.rx_buf.take().unwrap();
 
+                        let result = self.crc_check();
                         let frame_len = rbuf[MIMIC_PSDU_OFFSET as usize] as usize - radio::MFR_SIZE;
                         // Length is: S0 (0 Byte) + Length (1 Byte) + S1 (0 Bytes) + Payload
                         // And because the length field is directly read from the packet
                         // We need to add 2 to length to get the total length
 
-                        client.receive(rbuf, frame_len, self.registers.crcstatus.get() == 1, result)
+                        client.receive(
+                            rbuf,
+                            frame_len,
+                            self.registers.crcstatus.get() == 1,
+                            result,
+                        );
                     });
                 }
-                // Radio state - Disabled
-                _ => (),
             }
-            self.radio_off();
-            self.radio_initialize();
-            self.rx();
         }
+
+        // Enabling hardware shortcuts allows for a much faster operation. However, this can also lead
+        // to race conditions and strange edge cases. Namely, if a task_start or rx_en is set while interrupts are disabled,
+        // the event_end interrupt can be "missed" and the interrupt handler will not be called. If the event is missed, the state
+        // machine is unable to progress and the driver enters a deadlock
         self.enable_interrupts();
+        if rx_init {
+            self.rx();
+        }
+        if start_task {
+            self.registers.task_start.write(Task::ENABLE::SET);
+        }
     }
 
     pub fn enable_interrupts(&self) {
-        self.registers.intenset.write(
-            Interrupt::READY::SET
-                + Interrupt::CCAIDLE::SET
-                + Interrupt::CCABUSY::SET
-                + Interrupt::END::SET
-                + Interrupt::FRAMESTART::SET,
-        );
+        self.registers
+            .intenset
+            .write(Interrupt::READY::SET + Interrupt::CCABUSY::SET + Interrupt::END::SET);
```

Comment:
What is the reasoning behind this change? Seems odd that this just enables a different set of interrupts now

Reply:
The previous implementation enabled: 
`Interrupt::READY::SET`
`Interrupt::CCAIDLE::SET`
`Interrupt::CCABUSY::SET`
`Interrupt::END::SET`
`Interrupt::FRAMESTART::SET`

The framestart interrupt notifies the kernel of the progressing state of the rx/tx, but does not require any response or change of state to the radio so there does not exist a need to trigger an interrupt in the kernel for this radio event.

To reduce tx latency, I enabled the nrf52840 radio's hardware shortcuts feature so that the tx progression can automatically launch the next task on triggered radio interrupts. These shortcuts made it redundant to enable these interrupts as the internal radio events automatically trigger the next task and we no longer need to handle those events in the kernel.